### PR TITLE
Enabled parallel computing with PETSc in deformation analysis

### PR DESCRIPTION
**Pull request by wenqing**
as titled.

File SmallDeformationCommon/Common.h is removed, and its associated part is removed as well. With the functions in that file, the nodal force vector computation is repeated.

If the removed part is needed, commit 
 	[pDM] Corrected getNodalForces for parallel computing 			4289841
, which corrects the parallel nodal force computation,  will be selected.

A test with linear element is added to the Ctest list.

testparalleldm

### CODE REVIEW

**Comment by wenqing** [comment]
To resolve  issue #1765.
Future work: secondary variable output under MPI.

**Comment by wenqing** [comment]
This PR follows up PR #1797.

**Comment by endJunction** [comment]
Extending of mechanics processes for parallel runs is great. We could use some speedup along this lines.
Removing nodal force computation is not an option at all. There will be material forces added to the same interface.

**Comment by wenqing** [comment]
Added Common.h again. The file has been changed for parallel computing, however the functions in it are not used in this PR after commit  4289841.

File: ProcessLib/SmallDeformation/SmallDeformationFEM.h [hunk 1]
```diff
@@ -88,7 +88,6 @@ struct SecondaryData
 
 struct SmallDeformationLocalAssemblerInterface
     : public ProcessLib::LocalAssemblerInterface,
-      public ProcessLib::SmallDeformation::NodalForceCalculationInterface,
```
Comment (endJunction) [inline ProcessLib/SmallDeformation/SmallDeformationFEM.h]: Could you please explain why the interface is removed?

**Comment by endJunction** [comment]
I'm not sure about the changes to the computation of the nodal forces. When source terms would be added to the process (e.g. gravity) would this require changes to the nodal forces computation?

Independent of this, could you please drop the commit which is removing the "Common.h" file?

The independent calculation of the nodal forces allows an integration into other (small deformation) processes like HM. Now this become more difficult I think.

Let's discuss this before more time is invested in the changes.

**Comment by endJunction** [comment]
I have compared the provided `pvtu` result with the result of the serial run with vtkdiff, and it shows a rather large deviation:
```
Computed difference between data arrays:
abs l1 norm      = 6.393239226763097e+01
abs l2-norm^2    = 2.619448994702831e+00
abs l2-norm      = 1.618471190569307e+00
abs maximum norm = 1.105040679911180e-01
```

**Comment by wenqing** [comment]
@endJunction  From the point view of computational mechanics, which originated the FEM,  the discretized weak form terms of all source terms and also  coupled terms in the momentum balance equation are added to nodal force to have momentum balance numerically. Therefore using the RHS of the system of equation of deformation process as the nodal force is corrected except for the case of M being coupled to other processes and the monolithic scheme being applied.  Under such consideration, I think, we only need to computing the extra contribution to the nodal force from the coupled terms in the process classes of HM, TM or THM with monolithic scheme.

**Comment by wenqing** [comment]
or using the current nodal force computation in Common.h but add computation of total stress in HydroMechanicsFEM  and other similar assemblers.

File: ProcessLib/SmallDeformation/SmallDeformationProcess.h [hunk 2]
```diff
@@ -201,12 +202,12 @@ class SmallDeformationProcess final : public Process
...
-    void postTimestepConcreteProcess(GlobalVector const&) override
+    void postTimestepConcreteProcess(GlobalVector const& /*x*/) override
     {
         DBUG("PostTimestep SmallDeformationProcess.");
-
-        ProcessLib::SmallDeformation::writeNodalForces(
-            *_nodal_forces, _local_assemblers, *_local_to_global_index_map);
+        std::transform(_nodal_forces->begin(), _nodal_forces->end(),
+            _nodal_forces->begin(),
+            std::bind1st(std::multiplies<double>(), -1.0));
```
Comment (wenqing) [inline ProcessLib/SmallDeformation/SmallDeformationProcess.h]:
```c++
transform(n.b, n.e, o.b, [](double v) { return -v;});
```

File: ProcessLib/SmallDeformation/SmallDeformationProcess.h [hunk 3]
```diff
@@ -186,6 +185,10 @@ class SmallDeformationProcess final : public Process
             _global_assembler, &VectorMatrixAssembler::assembleWithJacobian,
             _local_assemblers, *_local_to_global_index_map, t, x, xdot,
             dxdot_dx, dx_dx, M, K, b, Jac, coupling_term);
+
+        b.copyValues(*_nodal_forces);
+        std::transform(_nodal_forces->begin(), _nodal_forces->end(),
+            _nodal_forces->begin(), [](double val) { return -val;});
```
Comment (wenqing) [inline ProcessLib/SmallDeformation/SmallDeformationProcess.h]: Setting negative values of ``_nodal_forces`` is moved here.

File: ProcessLib/SmallDeformation/SmallDeformationProcess.h [hunk 4]
```diff
@@ -201,14 +204,6 @@ class SmallDeformationProcess final : public Process
             _local_assemblers, *_local_to_global_index_map, x, t, dt);
     }
 
-    void postTimestepConcreteProcess(GlobalVector const&) override
```
Comment (wenqing) [inline ProcessLib/SmallDeformation/SmallDeformationProcess.h]: Removed this overloaded member.

File: ProcessLib/SmallDeformation/SmallDeformationProcess.h (hunk 2 above)
Comment (wenqing) [inline ProcessLib/SmallDeformation/SmallDeformationProcess.h]: Changed by amending the second commit.

**Comment by wenqing** [comment]
Data pointer has been updated.

**Pull request merged**